src/repo.rs

```diff
@@ -707,19 +735,25 @@ impl DocumentInfo {
         if !self.state.should_save() {
             return;
         }
-        let should_compact = self.patches_since_last_save > 10;
-        let storage_fut = if should_compact {
-            let to_save = {
-                let mut doc = self.document.write();
-                doc.automerge.save()
+        let should_compact =
+            self.changes_since_last_compact > self.allowable_changes_until_compaction;
+        let (storage_fut, new_heads) = if should_compact {
+            let (to_save, new_heads) = {
+                let doc = self.document.read();
+                (doc.automerge.save(), doc.automerge.get_heads())
             };
-            storage.compact(document_id.clone(), to_save)
+            self.changes_since_last_compact = 0;
+            (storage.compact(document_id.clone(), to_save), new_heads)
         } else {
-            let to_save = {
-                let mut doc = self.document.write();
-                doc.automerge.save_incremental()
+            let (to_save, new_heads) = {
+                let doc = self.document.read();
+                (
+                    doc.automerge.save_after(&self.last_heads),
+                    doc.automerge.get_heads(),
+                )
             };
-            storage.append(document_id.clone(), to_save)
+            self.changes_since_last_compact.checked_add(1).unwrap_or(0);
```

Comment:
Is this line no redundant given we count inside `note_changes`?

Reply:
If I have read the code correctly, the only time `save_document` is called without `note_changes` being called beforehand, is when the document is being closed / the repo is shutting down. So it shouldn't make any difference, right? I think it should be safe to remove this line.

Reply:
/cc @issackelly just in case we've missed something...

Reply:
> I think it should be safe to remove this line.

Agree, thanks for looking into it.